py/redrock/rebin.py

```diff
@@ -10,6 +10,169 @@
 import numpy as np
 import numba
 
+#Moved CUPY import to local within method.  Set global var to None here
+cp = None
+
+#Since CUPY import is moved, vars block_size and cuda_source will always be
+#set but this should cost no real time.
+block_size = 512 #Default block size, should work on all modern nVidia GPUs
+
+# cuda_source contains raw CUDA kernels to be loaded as CUPY module
+cuda_source = r'''
+        extern "C" {
+            __global__ void batch_trapz_rebin(const double* x, const double* y, const double* edges, const double* myz, double* result, int nz, int nbin, int nbasis, int nt) {
```

Comment:
Did you try implementing these with cupy or [numba for cuda](https://numba.readthedocs.io/en/stable/cuda/overview.html#overview) first?  Very few people within the collaboration know pure-cuda programming, so for maintainability it would be highly preferable to keep these in a higher-level more python-like form, and only resort to pure cuda kernels when absolutely necessary to speed up a function whose slowness is a blocking factor and isn't amenable to other GPUification.

Reply:
I have not done anything with numba for cuda, but it is not possible as far as I can tell in pure-python cupy as its not quite so simple as just performing a couple of vectorized operations.  I can look into numba for cuda but I suspect that will not work either.  However, as this is essentially a low level library operation, I would think the maintenance would be minimal.  I will let you know if after looking into it, I think numba for cuda is an option.